### ISSUE TITLE
Symlink read break out

### ISSUE THREAD

**Issue by Mickael-van-der-Beek** [yaakov123/hagana#3]
Found a new filesystem access control bypass.

You can simply create a symlink within the module's folder that points to a file outside of the `root` folder and read said symlink.

e.g:

```
import fs from 'fs';
import path from 'path';

export function run () {
  const srcPath = path.resolve('./my-symlink');
  const dstPath = '/some/folder/outside/root/mysecret.txt';
  fs.symlinkSync(dstPath, srcPath);
  const content = fs.readFileSync(srcPath, { encoding: 'utf8' });
  console.log('CONTENT=', content);
}
```

**Comment by Mickael-van-der-Beek** [comment]
Having though about this problem and its various solutions along the years I believe that you'll have troubles securing Hagana due to how it implements the access control.

Since you wrap or override existing Node.js APIs, you will always be chasing down lesser known APIs or OS behaviours that allow an attacker to get read / write access to the file system, network or shell.

I would probably advise you to consider a more syscall based approach that would always work, no matter which Node.js, V8 or libuv API is called.

TypeScript takes a similar approach to you I believe and has (or had until recently) the same symlink vulnerability.

**Comment by yaakov123** [comment]
Fixed in v0.0.15

**Comment by yaakov123** [comment]
> Sure, perfection doesn't exist. So 99% coverage is better than nothing. That's absolutely true.
However, if Hagana were to be used more widely, Hagana specific bypasses would be included in every supply chain attack just in case the consuming application were to use Hagana.

This is true, but my hope is that there won't be a generally available bypass for Hagana. It will take a determined attacked to find a bypass for Hagana and include it in their attack.
